README.md
status-area-applet-tor
======================

Description
-----------

A Hildon status area applet to control a (local) Tor instance.

Building
--------

You will need Hildon development libraries and headers. See
`configure.ac` for the exact dependencies.

```
autoreconf -i
./configure --disable-static
make && make install
```

Screenshots
-----------

TODO


debian/changelog
status-area-applet-tor (0.1) unstable; urgency=medium

  * Initial packaging

 -- Ivan J. <parazyd@dyne.org>  Fri, 27 Mar 2020 14:31:40 +0100


src/ind.c
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 * Copyricht (c) 2019 Christoph Lohmann <20h@r-36.net>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <limits.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ind.h"

void edie(char *fmt, ...)
{
	va_list fmtargs;

	va_start(fmtargs, fmt);
	vfprintf(stderr, fmt, fmtargs);
	va_end(fmtargs);
	fprintf(stderr, ": ");

	perror(NULL);

	exit(1);
}

void *reallocz(void *p, int l, int z)
{
	p = realloc(p, l);
	if (p == NULL)
		edie("realloc");
	if (z)
		memset(p, 0, l);

	return p;
}

void *mallocz(int l, int z)
{
	return reallocz(NULL, l, z);
}

void *memdups(void *p)
{
	char *ret;
	int l;

	l = strlen(p) + 1;
	ret = reallocz(NULL, l, 2);
	memmove(ret, p, l);

	return (void *)ret;
}

char *vsmprintf(char *fmt, va_list fmtargs, int size)
{
	char *ret;

	ret = reallocz(NULL, ++size, 2);
	vsnprintf(ret, size, fmt, fmtargs);

	return ret;
}

char *smprintf(char *fmt, ...)
{
	va_list fmtargs;
	char *ret;
	int len;

	va_start(fmtargs, fmt);
	len = vsnprintf(NULL, 0, fmt, fmtargs);
	va_end(fmtargs);

	if (len == INT_MAX)
		return NULL;

	va_start(fmtargs, fmt);
	ret = vsmprintf(fmt, fmtargs, len);
	va_end(fmtargs);

	return ret;
}


src/ind.h
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 * Copyright (c) 2019 Christoph Lohmann <20h@r-36.net>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef __IND_H__
#define __IND_H__

#define nelem(x) (sizeof(x) / sizeof((x)[0]))

#define MAXLINESIZE 1048576

void edie(char *fmt, ...);
void *reallocz(void *p, int l, int z);
void *mallocz(int l, int z);
void *memdups(void *p);
char *vsmprintf(char *fmt, va_list fmtargs, int size);
char *smprintf(char *fmt, ...);

#endif


src/tctl.c
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>

#include "ind.h"
#include "socket.h"
#include "tctl.h"

int tctl_check_error(char *err)
{
	if (err != NULL)
		/* TODO: Error handling */
		if (strncmp("250", err, 3))
			return 1;

	return 0;
}

void tctl_cmd(sock_t *sock, char *cmd, ...)
{
	va_list ap;
	char *req, *breq, *arg;

	req = smprintf("%s", cmd);

	va_start(ap, cmd);
	for (arg = va_arg(ap, char *); arg; arg = va_arg(ap, char *)) {
		breq = smprintf("%s %s", req, arg);
		free(req);
		req = breq;
	}
	va_end(ap);
	/* printf("%s\n", req); */
	sock_printf(sock, "%s\r\n", req);
	free(req);
}

void tctl_simplecmd(sock_t *sock, char *cmd)
{
	tctl_cmd(sock, cmd, NULL);
}

char *tctl_read_auth_cookie(char *path)
{
	int i, j;
	char cookie[65], *ret;
	unsigned char rbuf[32]; /* 32 is the current size of ControlAuthCookie */
	FILE* file = fopen(path, "rb");

	if (file == NULL) {
		return NULL;
		/* edie("tctl_read_auth_cookie"); */
	}

	fread(rbuf, 32, 1, file);
	fclose(file);

	for (i = 0, j = 0; i < 32; i++, j+=2)
		sprintf(cookie+j, "%02x", rbuf[i] & 0xff);

	ret = memdups(cookie);
	return ret;
}

/*
int main(void)
{
	sock_t *s;
	char *ck, *res;

	ck = tctl_read_auth_cookie("/var/lib/tor/auth_cookie");

	s = sock_new();
	if (sock_connect(s, "127.0.0.1", 9051))
		return 1;

	tctl_cmd(s, "authenticate", ck);
	free(ck);

	res = sock_gets(s);
	printf("got: %s", res);

	printf("%d\n", tctl_check_error(res));

	tctl_simplecmd(s, "getinfo network-liveness");
	res = sock_gets(s);
	printf("got: %s", res);
	printf("%d\n", tctl_check_error(res));

	if (sock_close(s) < 0) {
		free(s);
		return 1;
	}

	if (s != NULL)
		free(s);

	return 0;
}
*/


src/tctl.h
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef __TCTL_H__
#define __TCTL_H__

#include "socket.h"

int tctl_check_error(char *err);

void tctl_cmd(sock_t *sock, char *cmd, ...);
void tctl_simplecmd(sock_t *sock, char *cmd);

char *tctl_read_auth_cookie(char *path);

#endif


src/socket.c
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 * Copyright (c) 2019 Christoph Lohmann <20h@r-36.net>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <arpa/inet.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#include "ind.h"
#include "socket.h"

sock_t *sock_new()
{
	sock_t *ret;
	ret = mallocz(sizeof(sock_t), 2);
	return ret;
}

int sock_connect(sock_t *sock, char *server, unsigned short port)
{
	struct sockaddr_in remote;

	remote.sin_family = AF_INET;
	remote.sin_port = htons(port);
	remote.sin_addr.s_addr = inet_addr(server);

	sock->fd = socket(AF_INET, SOCK_STREAM, 0);
	if (sock->fd < 0)
		return 1;

	if (!connect(sock->fd, (struct sockaddr *)&remote, sizeof(struct sockaddr_in)))
		return 0;

	close(sock->fd);
	return 1;
}

int sock_close(sock_t *sock)
{
	return close(sock->fd);
}

int sock_write(sock_t *sock, char *buf, int len)
{
	return send(sock->fd, buf, len, MSG_NOSIGNAL);
}

int sock_writeall(sock_t *sock, char *buf, int len)
{
	int olen, nlen;

	for (olen = 0;
			(nlen = sock_write(sock, &buf[olen], len-olen))
			&& olen < len;
			olen += nlen);

	return 0;
}

int sock_printf(sock_t *sock, char *fmt, ...)
{
	va_list fmtargs;
	char *buf;
	int len;

	va_start(fmtargs, fmt);
	len = vsnprintf(NULL, 0, fmt, fmtargs);
	va_end(fmtargs);

	va_start(fmtargs, fmt);
	buf = vsmprintf(fmt, fmtargs, len);
	va_end(fmtargs);

	len = sock_writeall(sock, buf, strlen(buf));
	free(buf);
	return len;
}

int sock_read(sock_t *sock, char *buf, int len)
{
	return recv(sock->fd, buf, len, 0);
}

int sock_readall(sock_t *sock, char *buf, int len)
{
	int olen, nlen;

	for (olen = 0;
			(nlen = sock_read(sock, &buf[olen], len-olen))
			&& olen < len;
			olen += nlen);

	return 0;
}

char *sock_gets(sock_t *sock)
{
	char *ret;
	int len;

	len = 1;
	ret = mallocz(len, 2);
	while (sock_read(sock, &ret[len-1], 1) > 0 && ret[len-1] != '\n'
			&& len < MAXLINESIZE) {
		ret = reallocz(ret, ++len + 1, 0);
	}
	/*
	if (ret[len-1] != '\n') {
		free(ret);
		return NULL;
	}
	ret[len-1] = '\0';
	*/
	ret[len] = '\0';
	return ret;
}


src/socket.h
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 * Copyright (c) 2019 Christoph Lohmann <20h@r-36.net>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef __TORSOCKET_H__
#define __TORSOCKET_H__

typedef struct sock_t sock_t;
struct sock_t {
	int fd;
	void *data[2];
};

sock_t *sock_new();

int sock_connect(sock_t *sock, char *server, unsigned short port);
int sock_close(sock_t *sock);

int sock_write(sock_t *sock, char *buf, int len);
int sock_writeall(sock_t *sock, char *buf, int len);
int sock_printf(sock_t *sock, char *fmt, ...);
int sock_read(sock_t *sock, char *buf, int len);
int sock_readall(sock_t *sock, char *buf, int len);

char *sock_gets(sock_t *sock);

#endif


tor_transparent_proxy.sh
#!/bin/sh

iface="$(ip -4 r | grep '^default via' | sed 1q | cut -d' ' -f5)"

if [ -z "$iface" ]; then
	echo "Error: No network interface found." >&2
	exit 1
fi

enable_tproxy() {
	cp -f /etc/resolv.conf /etc/resolv.conf.orig
	echo "nameserver 127.0.0.1" > /etc/resolv.conf

	_tor_uid="$(id -u debian-tor)"
	_trans_port="9040"
	_dns_port="5353"
	_virt_addr="10.192.0.0/10"
	_out_if="$iface"

	# LAN destinations that shouldn't be routed through Tor
	_non_tor="127.0.0.0/8 10.0.0.0/8 172.16.0.0/12 192.168.0.0/16"

	# Other IANA reserved blocks (These are not processed by tor and dropped by default)
	_resv_iana="0.0.0.0/8 100.64.0.0/10 169.254.0.0/16 192.0.0.0/24 192.0.2.0/24 192.88.99.0/24 198.18.0.0/15 198.51.100.0/24 203.0.113.0/24 224.0.0.0/4 240.0.0.0/4 255.255.255.255/32"

	### Don't lock yourself out after the flush
	#iptables -P INPUT ACCEPT
	#iptables -P OUTPUT ACCEPT

	iptables -F
	iptables -t nat -F

	### *nat OUTPUT (For local redirection)
	# nat .onion addresses
	iptables -t nat -A OUTPUT -d $_virt_addr \
		-p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN \
		-j REDIRECT --to-ports $_trans_port

	# nat dns requests to Tor
	iptables -t nat -A OUTPUT -d 127.0.0.1/32 \
		-p udp -m udp --dport 53 -j REDIRECT --to-ports $_dns_port

	# Don't nat the Tor process, the loopback, or the local network
	iptables -t nat -A OUTPUT -m owner --uid-owner $_tor_uid -j RETURN
	iptables -t nat -A OUTPUT -o lo -j RETURN

	# Allow lan access for hosts in $_non_tor
	for _lan in $_non_tor; do
	  iptables -t nat -A OUTPUT -d $_lan -j RETURN
	done

	for _iana in $_resv_iana; do
	  iptables -t nat -A OUTPUT -d $_iana -j RETURN
	done

	# Redirect all other pre-routing and output to Tor's TransPort
	iptables -t nat -A OUTPUT -p tcp -m tcp \
		--tcp-flags FIN,SYN,RST,ACK SYN -j REDIRECT --to-ports $_trans_port

	### *filter INPUT
	# Don't forget to grant yourself ssh access from remote machines before the DROP.
	#iptables -A INPUT -i $_out_if -p tcp --dport 22 -m state --state NEW -j ACCEPT

	iptables -A INPUT -m state --state ESTABLISHED -j ACCEPT
	iptables -A INPUT -i lo -j ACCEPT

	# Allow INPUT from lan hosts in $_non_tor
	# Uncomment these 3 lines to enable.
	#for _lan in $_non_tor; do
	# iptables -A INPUT -s $_lan -j ACCEPT
	#done

	# Log & Drop everything else. Uncomment to enable logging
	#iptables -A INPUT -j LOG --log-prefix "Dropped INPUT packet: " --log-level 7 --log-uid
	iptables -A INPUT -j DROP

	### *filter FORWARD
	iptables -A FORWARD -j DROP

	### *filter OUTPUT
	iptables -A OUTPUT -m state --state INVALID -j DROP
	iptables -A OUTPUT -m state --state ESTABLISHED -j ACCEPT

	# Allow Tor process output
	iptables -A OUTPUT -o $_out_if -m owner --uid-owner $_tor_uid \
		-p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m state --state NEW -j ACCEPT

	# Allow loopback output
	iptables -A OUTPUT -d 127.0.0.1/32 -o lo -j ACCEPT

	# Tor transproxy magic
	iptables -A OUTPUT -d 127.0.0.1/32 -p tcp -m tcp --dport $_trans_port \
		--tcp-flags FIN,SYN,RST,ACK SYN -j ACCEPT

	# Allow OUTPUT to lan hosts in $_non_tor
	# Uncomment these 3 lines to enable.
	#for _lan in $_non_tor; do
	# iptables -A OUTPUT -d $_lan -j ACCEPT
	#done

	# Log & Drop everything else. Uncomment to enable logging
	#iptables -A OUTPUT -j LOG --log-prefix "Dropped OUTPUT packet: " --log-level 7 --log-uid
	iptables -A OUTPUT -j DROP

	### Set default policies to DROP
	iptables -P INPUT DROP
	iptables -P FORWARD DROP
	iptables -P OUTPUT DROP

	### Set default policies to DROP for IPv6
	ip6tables -P INPUT DROP
	ip6tables -P FORWARD DROP
	ip6tables -P OUTPUT DROP
}

disable_tproxy() {
	cp -f /etc/resolv.conf.orig /etc/resolv.conf

	iptables -F
	iptables -t nat -F
	iptables -P INPUT ACCEPT
	iptables -P FORWARD ACCEPT
	iptables -P OUTPUT ACCEPT
	ip6tables -P INPUT ACCEPT
	ip6tables -P FORWARD ACCEPT
	ip6tables -P OUTPUT ACCEPT
}

case "$1" in
enable)
	enable_tproxy
	;;
disable)
	disable_tproxy
	;;
esac


src/status-area-applet-tor.c
/*
 * Copyright (c) 2020 Ivan Jelincic <parazyd@dyne.org>
 * Copyright (c) 2015 Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>
 *
 * This file is part of status-area-applet-tor
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <gtk/gtk.h>
#include <hildon/hildon.h>
#include <libhildondesktop/libhildondesktop.h>

#include "tctl.h"
#include "socket.h"

#define STATUS_AREA_APPLET_TOR_TYPE (status_area_applet_tor_get_type())
#define STATUS_AREA_APPLET_TOR(obj) (G_TYPE_CHECK_INSTANCE_CAST ((obj), \
			STATUS_AREA_APPLET_TOR_TYPE, StatusAreaAppletTor))

typedef struct _StatusAreaAppletTor StatusAreaAppletTor;
typedef struct _StatusAreaAppletTorClass StatusAreaAppletTorClass;
typedef struct _StatusAreaAppletTorPrivate StatusAreaAppletTorPrivate;

struct _StatusAreaAppletTor
{
	HDStatusMenuItem parent;
	StatusAreaAppletTorPrivate *priv;
};

struct _StatusAreaAppletTorClass
{
	HDStatusMenuItemClass parent;
};

struct _StatusAreaAppletTorPrivate
{
	GtkWidget *conn_icon;
	GtkWidget *menu_button;
	GtkWidget *proxy_chkbutton;
	GtkWidget *settings_dialog;
	gboolean proxying_state;
	gint connection_state;
	gchar *auth_cookie;
	sock_t *sock;
};

HD_DEFINE_PLUGIN_MODULE_EXTENDED(StatusAreaAppletTor, status_area_applet_tor,
		HD_TYPE_STATUS_MENU_ITEM, G_ADD_PRIVATE_DYNAMIC(StatusAreaAppletTor),,);

static void status_area_applet_tor_set_status_icon(gpointer data, GdkPixbuf *pixbuf)
{
	StatusAreaAppletTor *item = STATUS_AREA_APPLET_TOR(data);
	g_return_if_fail(item != NULL && pixbuf != NULL);
	hd_status_plugin_item_set_status_area_icon(HD_STATUS_PLUGIN_ITEM(item), pixbuf);
	pixbuf = NULL;
}

static void status_area_applet_tor_set_menu_icon(StatusAreaAppletTor *self, GdkPixbuf *pixbuf)
{
	StatusAreaAppletTorPrivate *priv = status_area_applet_tor_get_instance_private(self);
	priv->conn_icon = gtk_image_new_from_pixbuf(pixbuf);
	hildon_button_set_image(HILDON_BUTTON(priv->menu_button), priv->conn_icon);
	pixbuf = NULL;
}

static void status_area_applet_tor_set_icons(StatusAreaAppletTor *self)
{
	StatusAreaAppletTorPrivate *priv = status_area_applet_tor_get_instance_private(self);
	GdkPixbuf *pixbuf = NULL;
	GtkIconTheme *theme = gtk_icon_theme_get_default();

	if (priv->connection_state)
		pixbuf = gtk_icon_theme_load_icon(theme, "statusarea_tor_enabled", 48, 0, NULL);
	else
		pixbuf = gtk_icon_theme_load_icon(theme, "statusarea_tor_disabled", 48, 0, NULL);
	status_area_applet_tor_set_menu_icon(self, pixbuf);

	if (priv->proxying_state) {
		pixbuf = gtk_icon_theme_load_icon(theme, "statusarea_tor_connected", 18, 0, NULL);
		status_area_applet_tor_set_status_icon(self, pixbuf);
	}

	g_object_unref(pixbuf);

	hildon_button_set_value(HILDON_BUTTON(priv->menu_button),
			priv->connection_state ? "Connected" : "Disconnected");
}

static void status_area_applet_tor_tctl_set_status(StatusAreaAppletTor *self)
{
	StatusAreaAppletTorPrivate *priv = status_area_applet_tor_get_instance_private(self);
	gchar *res;

	/* TODO: Refactor into multiple functions */

	if (sock_connect(priv->sock, "127.0.0.1", 9051)) {
		priv->connection_state = 0;
		return;
	}

	tctl_cmd(priv->sock, "authenticate", priv->auth_cookie);
	res = sock_gets(priv->sock);
	if (tctl_check_error(res))
		goto setstatusend;

	tctl_simplecmd(priv->sock, "getinfo network-liveness");
	res = sock_gets(priv->sock);
	if (tctl_check_error(res))
		goto setstatusend;

	if (!strncmp(res, "250-network-liveness=up", 23))
		priv->connection_state = 1;
	else
		priv->connection_state = 0;

setstatusend:
	sock_close(priv->sock);
	if (res != NULL)
		g_free(res);
	return;
}

static void status_area_applet_tor_proxy_chkbutton_toggled(GtkWidget *button,
		StatusAreaAppletTor *self)
{
	StatusAreaAppletTorPrivate *priv = status_area_applet_tor_get_instance_private(self);

	priv->proxying_state = !priv->proxying_state;

	if (priv->proxying_state)
		system("sudo /usr/sbin/tor_transparent_proxy.sh enable");
	else
		system("sudo /usr/sbin/tor_transparent_proxy.sh disable");
}

static void status_area_applet_tor_clicked(GtkWidget *button, StatusAreaAppletTor *self)
{
	StatusAreaAppletTorPrivate *priv = status_area_applet_tor_get_instance_private(self);

	GtkWidget *toplevel = gtk_widget_get_toplevel(button);
	gtk_widget_hide(toplevel);

	priv->settings_dialog = gtk_dialog_new();
	gtk_window_set_title(GTK_WINDOW(priv->settings_dialog), "Tor Status");
	gtk_window_set_transient_for(GTK_WINDOW(priv->settings_dialog), GTK_WINDOW(toplevel));
	gtk_window_set_destroy_with_parent(GTK_WINDOW(priv->settings_dialog), TRUE);

	priv->proxy_chkbutton = hildon_check_button_new(
			HILDON_SIZE_FINGER_HEIGHT|HILDON_SIZE_AUTO_WIDTH);
	gtk_button_set_label(GTK_BUTTON(priv->proxy_chkbutton), "Enable transparent system proxying");
	gtk_button_set_alignment(GTK_BUTTON(priv->proxy_chkbutton), 0.5f, 0.5f);
	hildon_check_button_set_active(HILDON_CHECK_BUTTON(priv->proxy_chkbutton),
			priv->proxying_state);
	gtk_box_pack_start(GTK_BOX(GTK_DIALOG(priv->settings_dialog)->vbox),
			priv->proxy_chkbutton, TRUE, TRUE, 0);

	g_signal_connect(priv->proxy_chkbutton, "toggled",
			G_CALLBACK(status_area_applet_tor_proxy_chkbutton_toggled), self);

	gtk_widget_show_all(priv->settings_dialog);
}

static void status_area_applet_tor_init(StatusAreaAppletTor *self)
{
	StatusAreaAppletTorPrivate *priv = status_area_applet_tor_get_instance_private(self);
	GdkPixbuf *pixbuf = NULL;
	GtkIconTheme *theme = gtk_icon_theme_get_default();

	self->priv = priv;
	priv->sock = sock_new();
	priv->connection_state = 0;
	priv->auth_cookie = tctl_read_auth_cookie("/var/run/tor/control.authcookie");
	priv->menu_button = hildon_button_new(HILDON_SIZE_FINGER_HEIGHT,
			HILDON_BUTTON_ARRANGEMENT_VERTICAL);

	status_area_applet_tor_tctl_set_status(self);
	status_area_applet_tor_set_icons(self);

	hildon_button_set_style(HILDON_BUTTON(priv->menu_button), HILDON_BUTTON_STYLE_PICKER);
	hildon_button_set_alignment(HILDON_BUTTON(priv->menu_button), 0.0, 0.0, 1.0, 1.0);
	hildon_button_set_title(HILDON_BUTTON(priv->menu_button), "Tor");

	g_signal_connect(priv->menu_button, "clicked",
			G_CALLBACK(status_area_applet_tor_clicked), self);

	gtk_container_add(GTK_CONTAINER(self), priv->menu_button);
	gtk_widget_show_all(GTK_WIDGET(self));
}

static void status_area_applet_tor_finalize(GObject *self)
{
	return;
}

static void status_area_applet_tor_class_init(StatusAreaAppletTorClass *class)
{
	GObjectClass *object_class = G_OBJECT_CLASS(class);
	object_class->finalize = status_area_applet_tor_finalize;
}

static void status_area_applet_tor_class_finalize(StatusAreaAppletTorClass *class)
{
	return;
}

StatusAreaAppletTor *status_area_applet_tor_new(void)
{
	return g_object_new(STATUS_AREA_APPLET_TOR_TYPE, NULL);
}
